adapters/httpapi_curl.c

```diff
@@ -294,9 +309,17 @@ static CURLcode ssl_ctx_callback(CURL *curl, void *ssl_ctx, void *userptr)
         HTTP_HANDLE_DATA *httpHandleData = (HTTP_HANDLE_DATA *)userptr;
 #ifdef USE_OPENSSL
         /*trying to set the x509 per device certificate*/
+        if (httpHandleData->x509privatekeytype == KEY_TYPE_ENGINE) {
+            ENGINE_load_builtin_engines();
+            httpHandleData->engine = ENGINE_by_id(httpHandleData->engineId);
+            if (httpHandleData->engine == NULL)
+            {
+                LogError("unable to load engine by ID: %s", httpHandleData->engineId);
+            }
+        }
         if (
             (httpHandleData->x509certificate != NULL) && (httpHandleData->x509privatekey != NULL) &&
-            (x509_openssl_add_credentials(ssl_ctx, httpHandleData->x509certificate, httpHandleData->x509privatekey, KEY_TYPE_DEFAULT, NULL) != 0)
+            (x509_openssl_add_credentials(ssl_ctx, httpHandleData->x509certificate, httpHandleData->x509privatekey, httpHandleData->x509privatekeytype, httpHandleData->engine) != 0)
            )
         {
             LogError("unable to x509_openssl_add_credentials");
```

Comment:
Here and everywhere below in the error handling cases, you would need to call `ENGINE_free` to avoid leaking the engine handle.

Reply:
I've added ENGINE_free now.